dm/pkg/checker/table_structure.go

```diff
@@ -79,68 +103,67 @@ func (c *TablesChecker) Check(ctx context.Context) *Result {
 		Name:  c.Name(),
 		Desc:  "check compatibility of table structure",
 		State: StateSuccess,
-		Extra: fmt.Sprintf("address of db instance - %s:%d", c.dbinfo.Host, c.dbinfo.Port),
 	}
 
-	var (
-		err        error
-		options    = make(map[string][]*incompatibilityOption)
-		statements = make(map[string]string)
-	)
-	for schema, tables := range c.tables {
-		if len(tables) == 0 {
-			tables, err = dbutil.GetTables(ctx, c.db, schema)
-			if err != nil {
-				markCheckError(r, err)
-				return r
-			}
-		}
+	startTime := time.Now()
 
-		for _, table := range tables {
-			tableName := dbutil.TableName(schema, table)
-			statement, err := dbutil.GetCreateTableSQL(ctx, c.db, schema, table)
-			if err != nil {
-				// continue if table was deleted when checking
-				if isMySQLError(err, mysql.ErrNoSuchTable) {
-					continue
-				}
-				markCheckError(r, err)
-				return r
-			}
+	for sourceID := range c.tableMap {
+		c.firstSourceID = sourceID
+		break
+	}
 
-			opts := c.checkCreateSQL(ctx, statement)
-			if len(opts) > 0 {
-				options[tableName] = opts
-				statements[tableName] = statement
-			}
+	checkCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	concurrency := maxThreadNum
+	for sourceID := range c.tableMap {
+		db, ok := c.dbs[sourceID]
+		if !ok {
+			markCheckError(r, errors.NotFoundf("client for sourceID %s", sourceID))
+			return r
 		}
+		maxConnections, err := utils.GetMaxConnections(ctx, db)
+		if err != nil {
+			markCheckError(r, err)
+			return r
+		}
+		concurrency = int(math.Min(float64(concurrency), float64((maxConnections+1)/2)))
 	}
 
-	for name, opts := range options {
-		if len(opts) == 0 {
-			continue
-		}
-		tableMsg := "table " + name + " "
+	for i := 0; i < concurrency; i++ {
+		c.checkWg.Add(1)
+		go c.checkTable(checkCtx, r)
+	}
 
-		for _, option := range opts {
-			switch option.state {
-			case StateWarning:
-				if r.State != StateFailure {
-					r.State = StateWarning
-				}
-				e := NewError(tableMsg + option.errMessage)
-				e.Severity = StateWarning
-				e.Instruction = option.instruction
-				r.Errors = append(r.Errors, e)
-			case StateFailure:
-				r.State = StateFailure
-				e := NewError(tableMsg + option.errMessage)
-				e.Instruction = option.instruction
-				r.Errors = append(r.Errors, e)
+outer:
+	for sourceID, tables := range c.tableMap {
+		for _, table := range tables {
+			select {
+			case c.inCh <- &checkItem{table, sourceID}:
+			case <-checkCtx.Done():
+				log.L().Logger.Warn("ctx canceled before input tables completely")
+				break outer
 			}
 		}
 	}
-
+	close(c.inCh)
+	c.checkWg.Wait()
+	close(c.errCh)
+outer2:
+	for {
+		select {
+		case err, ok := <-c.errCh:
+			if !ok {
+				break outer2
+			}
+			c.reMu.Lock()
+			markCheckError(r, err)
+			c.reMu.Unlock()
+		case <-checkCtx.Done():
+			break outer2
+		}
+	}
+	log.L().Logger.Info("check table structure over", zap.String("spend time", time.Since(startTime).String()))
 	return r
```

Comment:
how about extract to one or two functions for better review and test? e.g. func dispatchTables ...

Reply:
SGTM

Reply:
extract to two functions `dispatchCheckItem()` and `handleErr()` in https://github.com/pingcap/tiflow/pull/3975/commits/90442d12cf17c79db816a5cba7595a0cc08987cc